Add documentation on open endpoints/ports (#7277)
gardener/gardener: docs/usage/control-plane-endpoints-and-ports.md

# Endpoints and Ports of a Shoot Control-Plane

> With the reversed VPN tunnel, there are no endpoints with open ports in the shoot cluster required by gardener.

Reviewer: Please add a link to the reversed VPN usage documentation.

> In order to allow communication to the shoots control-plane in the seed cluster, there are shared endpoints for all shoots of a seed cluster.

Reviewer: In most cases, this might be true, but it is not true if exposure classes are used. Maybe we want to be more explicit. However, I am a bit undecided on this one.

Reviewer: On second thought: With multi-zonal seed clusters this is even less true.

Author: I wanted to keep it simple here. But you're right, oversimplification can make it wrong.

The main entry-point into the seed cluster is the loadbalancer of the Istio ingress-gateway service. Depending on the infrastructure provider, there can be one IP address per zone.

> The Istio ingress-gateway exposes the following TCP ports:

Reviewer: Actually, these are the load balancer ports and not the Istio ingress-gateway ports. At least the pods use partially different ports.

* **443** for requests to the shoot API-server. The request is dispatched according to the set TLS SNI extension.

> * **8443** for requests to the shoot API-server via `api-server-proxy`. 

Reviewer: You could add that these requests are dispatched based to the proxy protocol target, which is the IP address of `kubernetes.default.svc.cluster.local` in the shoot.

> * **8132** to establish the reversed VPN connection.

Reviewer: This is dispatched according to an http header value.

## Kube API-server via SNI 

> ![API-server via SNI](./images/api-server-sni.png)

Reviewer: We could add the port translation if we want. (The load balancer port is not equal to the istio pod port.)

Author: I added the it to the picture. However, I'm not sure if it's not beyond the intended scope and rather confusing.

DNS entries for `api.<external-domain>` and `api.<shoot>.<project>.<internal-domain>` point to the loadbalancer of an Istio ingress-gateway service.
The kubernetes client sets the server name to `api.<external-domain>` or `api.<shoot>.<project>.<internal-domain>`.
Based on SNI, the connection is forwarded to the respective API-server at TCP layer. There is no TLS termination at the Istio ingress-gateway.
TLS termination happens on the shoots API-server. Traffic is end-to-end encrypted between the client and the API-server. The certificate authority and authentication are defined in the corresponding `kubeconfig`.
Details can be found in [GEP-08](../proposals/08-shoot-apiserver-via-sni.md).

## Kube API-server via api-server-proxy

> ![API-server-proxy](./images/api-server-proxy.png)

Reviewer: Please adapt the picture:
`kubernetes.default.cluster.local` => `kubernetes.default.svc.cluster.local`
`kube-api-server.<shoot>.svc` => `kube-apiserver.<shoot>.svc`

Inside the shoot cluster, the API-server can also be reached by the cluster internal name `kubernetes.default.svc.cluster.local`. 

> The pods `api-server-proxy` are deployed in the host network as daemonset and intercepts connections to the kubernetes service IP address.

Reviewer: ```suggestion
The pods `apiserver-proxy` are deployed in the host network as daemonset and intercept connections to the kubernetes service IP address.
```

The destination address is changed to the cluster IP address of the service `kube-apiserver.<shoot-namespace>.svc.cluster.local` in the seed cluster.

> The connections are forwarded via an HTTP tunnel to the Istio ingress-gateway in the seed cluster.

Reviewer: This uses (HA-proxy's) proxy protocol. It is not an HTTP tunnel.

Reviewer: The proxy protocol is not encrypted. However, the end-to-end traffic itself is again encrypted the same way as with SNI. This is worth mentioning.

# Endpoints and Ports of a Shoot Control-Plane

With the [reversed VPN](./reversed-vpn-tunnel.md) tunnel, there are no endpoints with open ports in the shoot cluster required by gardener.
In order to allow communication to the shoots control-plane in the seed cluster, there are endpoints shared by multiple shoots of a seed cluster.

> Depending on the configured zones or [exposure classes](./exposureclasses.md) there are different endpoints in a seed cluster. The IP address can be determined by a DNS query for the API-server URL.

Reviewer: ```suggestion
Depending on the configured zones or [exposure classes](./exposureclasses.md) there are different endpoints in a seed cluster. The IP address(es) can be determined by a DNS query for the API-server URL.
```

> Depending on the configured zones or [exposure classes](./exposureclasses.md) there are different endpoints in a seed cluster. The IP address(es) can be determined by a DNS query for the API-server URL.

Reviewer 2: ```suggestion
Depending on the configured zones or [exposure classes](./exposureclasses.md), there are different endpoints in a seed cluster. The IP address(es) can be determined by a DNS query for the API-server URL.
```

> The main entry-point into the seed cluster is the load balancer of the Istio ingress-gateway service. Depending on the infrastructure provider, there can be one IP address per zone.

Reviewer: ```suggestion
The main entry-point into the seed cluster is the load balancer of the Istio ingress-gateway service. Depending on the infrastructure provider, there can be one IP address per availability zone.
```

The load balancer of the Istio ingress-gateway service exposes the following TCP ports:

* **443** for requests to the shoot API-server. The request is dispatched according to the set TLS SNI extension.
* **8443** for requests to the shoot API-server via `api-server-proxy`, dispatched based on the proxy protocol target, which is the IP address of kubernetes.default.svc.cluster.local in the shoot.

> * **8132** to establish the reversed VPN connection. It's dispatched according an HTTP header value.

Reviewer 2: ```suggestion
* **8132** to establish the reversed VPN connection. It's dispatched according to an HTTP header value.
```

> ## Kube API-server via SNI 

Reviewer 2: ```suggestion
## `kube-apiserver` via SNI 
```

![API-server via SNI](./images/api-server-sni.png)

DNS entries for `api.<external-domain>` and `api.<shoot>.<project>.<internal-domain>` point to the load balancer of an Istio ingress-gateway service.
The kubernetes client sets the server name to `api.<external-domain>` or `api.<shoot>.<project>.<internal-domain>`.
Based on SNI, the connection is forwarded to the respective API-server at TCP layer. There is no TLS termination at the Istio ingress-gateway.
TLS termination happens on the shoots API-server. Traffic is end-to-end encrypted between the client and the API-server. The certificate authority and authentication are defined in the corresponding `kubeconfig`.
Details can be found in [GEP-08](../proposals/08-shoot-apiserver-via-sni.md).

## Kube API-server via api-server-proxy

![API-server-proxy](./images/api-server-proxy.png)

> Inside the shoot cluster, the API-server can also be reached by the cluster internal name `kubernetes.default.svc.cluster.local`. 

Reviewer 2: Replace "API-server" with "API server" everywhere

> The pods `api-serverproxy` are deployed in the host network as daemonset and intercept connections to the kubernetes service IP address.

Reviewer 2: ```suggestion
The pods `apiserver-proxy` are deployed in the host network as daemonset and intercept connections to the Kubernetes service IP address.
```
?
